resource/ui/ts/models/timeseries.ts

```diff
@@ -37,37 +39,74 @@ module Models {
+        }
 
         /**
-         * Query wraps the information needed to make a single time series query
-         * to the server.
+         * Query dispatches a single time series query to the server.
+         */
+        function query(start:Date, end:Date, series:string[]):promise<QueryResultSet> {
+            var url = "/ts/query";
+            var data = {
+                start_nanos: start.getTime() * 1.0e6,
+                end_nanos: end.getTime() * 1.0e6,
+                queries: series.map((r) => {return {name: r};}),
+            }
+        
+            return m.request({url:url, method:"POST", extract:nonJsonErrors, data:data})
+                .then((d:QueryResultSet) => {
+                    // Populate missing collection fields with empty arrays.
+                    if (!d.results) {
+                        d.results = [];
+                    }
+                    d.results.forEach((r) => {
+                        if (!r.datapoints) {
+                            r.datapoints = []
+                        }
+                    });
+                    return d;
+                });
+        }
+
+        /** 
+         * StaticQuery is a time series query with a constant start and end
+         * time.
          */
-        export class Query {
-            private series:string[]
+        export class StaticQuery {
+            private series:string[];
+            private data:promise<QueryResultSet>;
             constructor(public start:Date, public end:Date, ...series:string[]) {
-                this.series = series
+                this.series = series;
             }
 
-            query():_mithril.MithrilPromise<QueryResultSet> {
-                var url = "/ts/query";
-                var data = {
-                    start_nanos: this.start.getTime() * 1.0e6,
-                    end_nanos: this.end.getTime() * 1.0e6,
-                    queries: this.series.map((r) => {return {name: r};}),
+            query():promise<QueryResultSet> {
+                if (this.data != null) {
+                    return this.data;
                 }
-                return m.request({url:url, method:"POST", extract:nonJsonErrors, data:data})
-                    .then((d:QueryResultSet) => {
-                        // Populate missing collection fields with empty arrays.
-                        if (!d.results) {
-                            d.results = [];
-                        }
-                        d.results.forEach((r) => {
-                            if (!r.datapoints) {
-                                r.datapoints = []
-                            }
-                        });
-                        return d;
-                    });
+                this.data = query(this.start, this.end, this.series);
+                return this.data
+            }
+        }
+
+        /** 
+         * SlidingQuery is a query which monitors a duration of constant size
```

Comment:
How does this actually implement the sliding query? I'm really confused by this javascript code. It doesn't appear to be any different from the StaticQuery except it returns the data directly instead of storing it.